src/sage/matrix/matrix2.pyx

```diff
@@ -16615,10 +16615,14 @@ cdef class Matrix(Matrix1):
         d = self.smith_form(transformation=False)
         r = min(self.nrows(), self.ncols())
         ed = [d[i, i] for i in range(r)]
-        try:
-            return [x.canonical_associate()[0] for x in ed]
-        except (AttributeError, TypeError):
-            return ed
+        # Normalize elements using canonical_associate for number field orders
+        from sage.rings.number_field.order import Order
+        if isinstance(self.base_ring(), Order):
```

Comment:
Is this intended to fix a bug? Which one?

I'm pretty sure I added this specifically for the case of integers, not `Order` elements. When **passagemath-modules** is tested, the specialized implementation of integer matrices is not available, and then the elementary divisors often turned out to be negative integers, which broke various code.